### PR TITLE
[MIG][9.0]stock analytic

### PR DESCRIPTION
**Pull request by AaronHForgeFlow**
Stock Analytic
==============

Adds an analytic account in stock move to be able to get analytic information
when generating the journal items.

Usage
=====

Assigned Stock Move With Analytic Account
-----------------------------------------

You need to:

#. Create manually or open draft picking
#. Add or edit stock move. Fill **analytic account** field

I closed the previous PR as I accidentally deleted the history: https://github.com/OCA/account-analytic/pull/81
PR to v8 merged: https://github.com/OCA/account-analytic/pull/53

### CODE REVIEW

**Comment by AaronHForgeFlow** [comment]
cc @eficent @darshan-serpent

**Comment by LoisRForgeFlow** [comment]
seems that runbot is not working for this repo, all v9 and v10 builds are skipped. Is there something we can do?

**Comment by serpentcs-dev1** [comment]
Code LGTM, functional testing pending.

File: stock_analytic/tests/test_stock_picking.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+        picking = self.env['stock.picking'].create(picking_data)
+
+        move_data = {
+            'picking_id': picking.id,
+            'product_id': self.product.id,
+            'location_id': location_id.id,
+            'location_dest_id': location_dest_id.id,
+            'date': datetime.now().strftime('%Y-%m-%d %H:%M:%S'),
+            'date_expected': datetime.now().strftime('%Y-%m-%d %H:%M:%S'),
+            'invoice_state': 'none',
```
Comment (serpentcs-dev1) [inline stock_analytic/tests/test_stock_picking.py]: invoice_state selection field no longer exists in stock moves for v9.

**Comment by AaronHForgeFlow** [comment]
@lreficent @darshan-serpent runbot is now available. All feedback is welcome :heart_eyes:

File: stock_analytic/README.rst [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+==============
+
+Adds an analytic account in stock move to be able to get analytic information
+when generating the journal items.
+
+
+Usage
+=====
+
+Assigned Stock Move With Analytic Account
```
Comment (serpentcs-dev1) [inline stock_analytic/README.rst]: Assigns

File: stock_analytic/models/stock.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    @api.model
+    def _prepare_account_move_line(self, move, qty, cost,
+                                   credit_account_id, debit_account_id,
+                                   context=None):
+        res = super(StockQuant, self)._prepare_account_move_line(
+            move, qty, cost, credit_account_id, debit_account_id,
+            context=context
+        )
+
+        # Add analytic account in debit line
```
Comment (serpentcs-dev1) [inline stock_analytic/models/stock.py]: Does the analytic account is just add to the debit line?

File: stock_analytic/__openerp__.py [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+    "depends": [
+        "stock_account",
+        "analytic",
+    ],
+    "data": [
+        "views/stock_move_views.xml",
+        "views/stock_inventory_views.xml",
+    ],
+    'installable': True,
+    'active': False,
```
Comment (serpentcs-dev1) [inline stock_analytic/__openerp__.py]: this is deprecated

**Review by serpentcs-dev1 — APPROVED** [review]
Reviewed code and tested functionality in local

Minor changes, Otherwise LGTM :+1:

File: stock_analytic/models/stock.py (hunk 3 above)
Comment (AaronHForgeFlow) [reply stock_analytic/models/stock.py]: Yes, exactly, just in the debit side. When receiving items we debit inventory accounts and when we delivery for a project we debit Expenses.

**Comment by AaronHForgeFlow** [comment]
@darshan-serpent your comments have been attended can you review again please?
@eficent

File: stock_analytic/models/stock.py (hunk 3 above)
Comment (serpentcs-dev1) [reply stock_analytic/models/stock.py]: I was wondering then why were you using `range(0, 2)` won't it update the account for [both the lines?!](https://github.com/odoo/odoo/blob/9.0/addons/stock_account/stock_account.py#L354)

File: stock_analytic/__openerp__.py [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+# Copyright 2015 Clear Corp
+# Copyright 2016 OpenSynergy Indonesia
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+{
+    "name": "Stock Analytic",
+    "summary": "Adds an analytic account in stock move",
+    "version": "9.0.1.0.0",
+    "author": "Julius Network Solutions,"
+    "ClearCorp, OpenSynergy Indonesia, "
+    "Odoo Community Association (OCA)",
```
Comment (LoisRForgeFlow) [inline stock_analytic/__openerp__.py]: Please indent this to enhance readability

File: stock_analytic/models/stock.py [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+
+class StockQuant(models.Model):
+
+    _inherit = "stock.quant"
+
+    @api.model
+    def _prepare_account_move_line(self, move, qty, cost,
+                                   credit_account_id, debit_account_id,
+                                   context=None):
```
Comment (LoisRForgeFlow) [inline stock_analytic/models/stock.py]: In new API you have context available in self. Fix this accordingly.

File: stock_analytic/views/stock_inventory_views.xml [hunk 7]
```diff
@@ -0,0 +1,15 @@
+<?xml version="1.0" encoding="utf-8"?>
+<openerp>
+    <data>
```
Comment (LoisRForgeFlow) [inline stock_analytic/views/stock_inventory_views.xml]: just `<odoo>`

Comment (LoisRForgeFlow) [reply stock_analytic/views/stock_inventory_views.xml]: Same for the rest of data files.

**Review by LoisRForgeFlow — COMMENTED** [review]
Some comments

**Comment by AaronHForgeFlow** [comment]
thank you for your review @lreficent I have attended your remarks :relaxed:

File: stock_analytic/models/stock.py (hunk 6 above)
Comment (LoisRForgeFlow) [reply stock_analytic/models/stock.py]: you hava to remove it here also! not only in super.

**Comment by LoisRForgeFlow** [comment]
@aheficent check https://github.com/OCA/account-analytic/pull/82#discussion_r121400608 !

File: stock_analytic/models/stock.py (hunk 6 above)
Comment (AaronHForgeFlow) [reply stock_analytic/models/stock.py]: oops :sweat_smile:

**Comment by AaronHForgeFlow** [comment]
I think now is :ok_hand:

**Review by LoisRForgeFlow — APPROVED** [review]
Code review -> LGTM :+1:

File: stock_analytic/models/stock.py [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+    @api.model
+    def _prepare_account_move_line(self, move, qty, cost,
+                                   credit_account_id, debit_account_id):
+        res = super(StockQuant, self)._prepare_account_move_line(
+            move, qty, cost, credit_account_id, debit_account_id)
+        # Add analytic account in debit line
+        if not move.analytic_account_id:
+            return res
+
+        for num in range(0, 2):
```
Comment (serpentcs-dev1) [inline stock_analytic/models/stock.py:32]: Why are you using `range(0, 2)` won't it update the account for both the [lines?!](https://github.com/odoo/odoo/blob/9.0/addons/stock_account/stock_account.py#L354)

Comment (AaronHForgeFlow) [reply stock_analytic/models/stock.py:32]: If there's a product in the credit line yes, but it does not happen normally.

**Comment by AaronHForgeFlow** [comment]
@elicoidal can you review this again? It's the same code you approved in this one: https://github.com/OCA/account-analytic/pull/81

**Review by elicoidal — APPROVED** [review]

**Comment by AaronHForgeFlow** [comment]
Thank you! :tada:

**Pull request merged**